<font color= 'darkred' size = '5'> **Python Mathematics** </font><br>

Python offers the math module to carry out different mathematics like trigonometry, logarithms, probability and statistics, etc. For example,

In [14]:
import math

print(math.pi)
print(f'{math.pi :.3f}')

print(math.cos(math.pi))

print(math.exp(10))

print(math.log10(1000))

print(math.sinh(1))

print(math.sqrt(5))

print(math.factorial(6))

print(math.fmod(9,4)) # 9%4



3.141592653589793
3.142
-1.0
22026.465794806718
3.0
1.1752011936438014
2.23606797749979
720
1.0


|function | description|
| :-      |    :-      | 
|ceil (x) | Returns the smallest integer greater than or equal to x|
| copysign(x, y)	|Returns x with the sign of y|
| fabs(x)	|Returns the absolute value of x (float)
|factorial(x)	|Returns the factorial of x
|floor(x)	|Returns the largest integer less than or equal to x
|fmod(x, y)	|Returns the remainder when x is divided by y
|frexp(x)	|Returns the mantissa and exponent of x as the pair (m, e)
|fsum(iterable)	|Returns an accurate floating point sum of values in the iterable
|ldexp(x, i)	|Returns x * (2**i)
|modf(x)	|Returns the fractional and integer parts of x
|trunc(x)	|Returns the truncated integer value of x
|exp(x)	|Returns e**x
|expm1(x)	|Returns e**x - 1
|log(x[, b])	|Returns the logarithm of x to the base b (defaults to e)
|log1p(x)	|Returns the natural logarithm of 1+x
|log2(x)	|Returns the base-2 logarithm of x
|log10(x)	|Returns the base-10 logarithm of x
|pow(x, y)	|Returns x raised to the power y
|sqrt(x)	|Returns the square root of x
|acos(x)	|Returns the arc cosine of x
|acosh(x)	|Returns the inverse hyperbolic cosine of x
|cos(x)	|Returns the cosine of x
|degrees(x)	|Converts angle x from radians to degrees
|radians(x)	|converts angle x from degrees to radians
|erf(x)	|Returns the error function at x
|gamma(x)	|Returns the Gamma function at x
|pi	|Returns (3.14159...)
|e	|mathematical constant e (2.71828...)|


<font color = 'steelblue' size = '4'>**Python Random Module**</font>

Python offers the random module to generate random numbers or to pick a random item from an iterator.

First we need to import the random module.

<font color = 'steelblue' size = 3>**The seed ( )**</font> 
>
>The seed method is used to initialize the random number generator.
>
>The random number generator needs a number to start with (a seed value), to be able to generate a random number.
>
>By default the random number generator uses the current system time.
>
>Use the seed() method to customize the start number of the random number generator.
>
>if u seed(10) you always start with 0.5714.. and if u use same seed twice u will get same random number

In [16]:
import random

random.seed(10)
print(random.random())

0.5714025946899135


<font color= 'steelblue' size = '3'>**The random ( ) randrange ( ) and randint ( )**</font>

The `random()` method returns a random floating number between 0 and 1.

The `randrange()` method returns a randomly selected element from the specified range.

>`random.randrange(start, stop, step)`

The `randint()` method returns an integer number selected element from the specified range.

>Note: This method is an alias for randrange(start, stop+1) / `random.randint(start, stop)`

<font color = 'steelblue' size = '3'>**The choice ( ) shuffle ( ) and sample ( )**</font>

>The `choice` method returns a randomly selected element from the specified sequence.
>
>The sequence can be a string, a range, a list, a tuple or any other kind of sequence
>
>the `shuffle` method changes the original list, it does not return a new list.
>
>The `sample()` method returns a list with a randomly selection of a specified number of items `k` from a sequnce. 

In [3]:
import random

listt = ['1','2','cup', 'coffee']

print(random.random(),' :random.random()')

print(random.randint(0,10), ':random.randint()') #including 10

print(random.randrange(0,12), ':random.randrange()') #not including 12

print(random.choice(listt), ' :a choice from my list')  # choice([1,5]) 1 ya 5 na baze 1 ta 5
random.shuffle(listt) #bor mizane
print(listt,' :shuffled list')

print(random.sample((listt), k=2), ' :2 random sampled list')



0.4288890546751146  :random.random()
9 :random.randint()
0 :random.randrange()
2  :a choice from my list
['1', 'cup', '2', 'coffee']  :shuffled list
['cup', '1']  :2 random sampled list


In [41]:
import datetime 
now = datetime.datetime.now()
print(now)

import sys
print(sys.version)
print(sys.platform)

import platform
print(platform.release())

import os 
os.getcwd()

2025-02-28 12:41:21.296242
3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:20:11) [MSC v.1938 64 bit (AMD64)]
win32
11


'C:\\Users\\zahra\\jupyters file'

<font color = 'steelblue' size = '3'>**The uniform() and gauss**</font>
    
>The `uniform` method returns a random floating number between the two specified numbers (both included).
>    
>`random.uniform(a, b)`
>
>The `gauss()` Returns a random float number based on the Gaussian distribution (used in probability theories)

###### <font color = 'darkred' size ='5'>**List**</font>

> <font color = 'darkred' size = '4'>**Usefull python list method**</font>
|Method |      description|
| :-    |          :-     |
|append()	|add an item to the end of the list|
|extend()	|add all the items of an iterable to the end of the list
|insert()	|inserts an item at the specified index
|remove()	|removes item present at the given index
|pop()	|returns and removes item present at the given index
|clear()	|removes all items from the list
|index()	|returns the index of the first matched item
|count()	|returns the count of the specified item in the list
|sort()	|sort the list in ascending/descending order
|reverse()	|reverses the item of the list
|copy()	|returns the shallow copy of the list|

<font color= 'steelblue' size = '4'>**Slicing of a List**</font>

In Python, it is possible to access a portion of a list using the slicing operator.



In [65]:
my_list = ['p','r','o','g','r','a','m','i','z']
print(len(my_list))
print(my_list.index('o'))   #or
my_list[5:10] = [1, 2, 3, 4, 5]
print(my_list)
print(6 in my_list)

print(my_list[2:5])
print(my_list[::-1])
print('______________________')

a = [1, 2, 3, 4]
b = ['1','2','a', [905,690]]
c = a + b
print(c)
print(b[3][1])
print('______________________')

a = [7, 5, 30, 2, 6 ,25]
print(max(a)) #or 

m = -1
for i in a:
    if i> m:
        m =i
print(m)

print(min(a))
print(sum(a))
print('______________________')

d = [ 1, 2, 3, 4, 2, 1, 5, 4, 3, 3, 3, 7, 1]
print(d.count(2))


9
2
['p', 'r', 'o', 'g', 'r', 1, 2, 3, 4, 5]
False
['o', 'g', 'r']
[5, 4, 3, 2, 1, 'r', 'g', 'o', 'r', 'p']
______________________
[1, 2, 3, 4, '1', '2', 'a', [905, 690]]
690
______________________
30
30
2
75
______________________
2


<font color = 'steelblue' siz= '4'> **Add Elements to a List**</font>


Lists are mutable (changeable). Meaning we can add and remove elements from a list.

Python list provides different methods to add items to a list.

1. **Using append()**

The append() method adds an (1) item at the end of the list.

2. **using insert()**

We use the insert() method to add an element at the specified index.

3. **using extend list()**

To append elements from another list to the current list, use the extend() method.<br>

The **extend()** method does not have to append **lists**, you can add any iterable object **(tuples, sets, dictionaries** etc.

In [75]:
numbers = [21, 34, 54, 12]
numbers.insert(2,'apple')
print(numbers)

numbers.append(32)
print("After Append:", numbers)

[21, 34, 'apple', 54, 12]
After Append: [21, 34, 'apple', 54, 12, 32]


In [97]:
numbers = [31,23,45,32, 2, 12, 14]
names= ['zara', 'neda', 'kave', 'navid']


numbers.extend(names)  #+
print(numbers)
print(len(numbers))

#lets see the difference between append and extend 

names.append(numbers)
print(names)
print(len(names))

numbers.remove(2)
print(numbers)
names.reverse()
print(names)

[31, 23, 45, 32, 2, 12, 14, 'zara', 'neda', 'kave', 'navid']
11
['zara', 'neda', 'kave', 'navid', [31, 23, 45, 32, 2, 12, 14, 'zara', 'neda', 'kave', 'navid']]
5
[31, 23, 45, 32, 12, 14, 'zara', 'neda', 'kave', 'navid']
[[31, 23, 45, 32, 12, 14, 'zara', 'neda', 'kave', 'navid'], 'navid', 'kave', 'neda', 'zara']


<font  color = 'steelblue' size = '3'> **an example of append**</font>


In [65]:
listt = []
n = 5

# you have 5 chance        
for i in range(0, 5):
    elements= int(input('enter your list of number: '))
    if 0<elements<100:
        listt.append(elements)
    else:
        print('wrong number!')
        elements= int(input('enter your list of number: '))
        listt.append(elements)
        
print (listt)

# there is A according to ur chance!
import random
a =random.randint(0,100)
print(a)

def winers_check():
    if a in listt:
        print('YOU WIN!!', 'your gift is 1 mil $')
    else:
        gift = 0
        print ('bad luck ha!? your gift is', gift, '$')
        
winers_check()
               

enter your list of number:  100


wrong number!


enter your list of number:  2
enter your list of number:  23
enter your list of number:  45
enter your list of number:  67
enter your list of number:  89


[2, 23, 45, 67, 89]
45
YOU WIN!! your gift is 1 mil $


In [142]:
a = []
for i in range (4):
    a.append(i)
print(a)
#or
a = [i*2 for i in range(4)]
print(a)

a = [i**2 for i in range(3,6)]
print(a)

a = [1, -2, 3, -5, 6, -10]
b = [abs(i) for i in a]
print(b)

import math 
a = [round(math.pi, i) for i in range(1, 5)]
print(a)

a = ['$sara', 'reza$']
b = [i.strip('$') for i in a]
print(b)

scores = [11,10,12,13,14,4,7,12,19]
b = [i for i in scores if i <10]
print(b)


[0, 1, 2, 3]
[0, 2, 4, 6]
[9, 16, 25]
[1, 2, 3, 5, 6, 10]
[3.1, 3.14, 3.142, 3.1416]
['sara', 'reza']
[4, 7]


In [13]:
import random
class_list = ['yazdani','abdi', 'zakeri', 'imani', 'fooladvand']
a = random.choice(class_list)
print('{} is the one!'.format(a))

yazdani is the one!


<font color = 'steelblue' size = '3'> **Remove an Item From a List**</font>

1. Using **del** Statement and **clear()**

In Python we can use the del statement to remove one or more items from a list, **clear**  will clean The list still remains, but it has no content.

2. Using **remove()**

We can also use the remove() method to delete a list item.

3. using **pop()**

The pop() method removes the specified index.





In [81]:
languages = ['Python', 'Swift', 'C++', 'C', 'Java', 'Rust', 'R', 'php', 'go']

del languages[0 : 2]  
print(languages)

languages.remove('C')
print(languages)

languages.pop(4) # remove the 4th index  pop() will remove the last one 
print(languages)

# delete entire list and error
print(languages.clear())

['C++', 'C', 'Java', 'Rust', 'R', 'php', 'go']
['C++', 'Java', 'Rust', 'R', 'php', 'go']
['C++', 'Java', 'Rust', 'R', 'go']
None


In [9]:
a = [1 ,2]
b = [1, 2, 3, 4,]
c=[]
for i in a:
    for j in b:
        c.append((i,j))
print(c)

print('___________________________')

a = [2.6, float('NaN'), 4.5, 6.3]    #removing nan item
B =[]
import math
for i in a:
    if not math.isnan(i):
        B.append(i)
print(B)


[(1, 1), (1, 2), (1, 3), (1, 4), (2, 1), (2, 2), (2, 3), (2, 4)]
___________________________
[2.6, 4.5, 6.3]


In [59]:
m = [             #matrix
    [1,2,3],
    [4,5,6],
    [7,8,9]
     ]
print(len(m))
print(m[0])
print('__________________')
for i in m:
    print(i)
    
for i in m:   #sootun
    print(i[0], end= '  ')
    
print()

for i in range(0,3):
    print(m[i][i], end= '  ')  #anasore qotri

print()

for i in range(0,3):         #qotre farE (i = 0,1,2) --> m[0][2], m[1][1], m[2][0]
    print(m[i][2-i], end= '  ')
print()    
print('_______________')

a = []
a.extend([sum(i) for i in m])    #jame satr ha
print(a)

print('_______________')

b = [] 
for col in range(3):                    #jame sotunha
    b.append(sum(i[col] for i in m))
print(b)

3
[1, 2, 3]
__________________
[1, 2, 3]
[4, 5, 6]
[7, 8, 9]
1  4  7  
1  5  9  
3  5  7  
_______________
[6, 15, 24]


In [27]:
fruits = ["apple", "banana", "cherry", "kiwi", "mango"]
newlist = []

for x in fruits:
    if "a" in x:
        newlist.append(x)
print(newlist)

#list comprehension


fruits = ["apple", "banana", "cherry", "kiwi", "mango"]

newlist = [x for x in fruits if "a" in x]

print(newlist)

['apple', 'banana', 'mango']
['apple', 'banana', 'mango']


**Condition** (list compehension)

The condition is like a **filter** that only accepts the items that valuate to True.

In [35]:
fruits = ["apple", "banana", "cherry", "kiwi", "mango"]

newlist = [x.upper() for x in fruits if x != "apple"]
newlist2 = [x if x != "banana" else "orange" for x in fruits]
print(newlist)
print(newlist2)

['BANANA', 'CHERRY', 'KIWI', 'MANGO']
['apple', 'orange', 'cherry', 'kiwi', 'mango']


<font color= 'steelblue' size = '3'> **sort**</font>

notice that sort is case insensitive, u can first lower them, then use sort

In [16]:
anylist = ['reza', 'Tara', 'Zari','ali']

anylist.sort()
print(anylist, 'sort befor lowercase')
anylist.sort(key = str.lower)
print(anylist,'sort after lower case')

numlist = [21323,23124,23432,23213,12421,23123,23412]
numlist.sort()
print(numlist)

mylist = ['reza', 'farhad', 'navid', 'hossien', 'kaveh','ali', 'younes'] 
mylist.sort(reverse = True) #barax but uppercase is important
print(mylist)

mylist.reverse() # regardless to alphabet it will reverse the list
print(mylist)


['Tara', 'Zari', 'ali', 'reza'] sort befor lowercase
['ali', 'reza', 'Tara', 'Zari'] sort after lower case
[12421, 21323, 23123, 23124, 23213, 23412, 23432]
['younes', 'reza', 'navid', 'kaveh', 'hossien', 'farhad', 'ali']
['ali', 'farhad', 'hossien', 'kaveh', 'navid', 'reza', 'younes']


<font color= 'steelblue' size = '3'>**Customize Sort Function**</font>

You can also customize your own function by using the keyword argument key = function.

The function will return a number that will be used to sort the list (the lowest number first):

In [28]:
def closerto(n):
    return abs (n - x)

x = int(input('your age: '))
this_list = [23,25,26,34,21,33,41,36,38]
this_list.sort(key = closerto)
print(this_list)

your age: 32
[33, 34, 36, 26, 38, 25, 23, 41, 21]


<font color = 'steelblue' size = '3'>**Copy a List**</font>

You cannot copy a list simply by typing **list2 = list1**, because: list2 will only <br>be a reference to list1, and changes made in list1 will automatically also be made in list2.

There are ways to make a copy, one way is to use the built-in List method **copy()**.

In [23]:
thislist = ["apple", "banana", "cherry"]
mylist = thislist.copy()
print(mylist) #or
mylist2 = list(thislist)
print(mylist2)

['apple', 'banana', 'cherry']
['apple', 'banana', 'cherry']


<font color = 'steelblue' sie = '3'>**join(), append(), extend()**</font>

In [27]:
list1 = ["a", "b" , "c"]
list2 = [1, 2, 3]
for x in list2:
    list1.append(x)
print(list1)

list1.extend(list2)
print(list1)

['a', 'b', 'c', 1, 2, 3]
['a', 'b', 'c', 1, 2, 3, 1, 2, 3]


<font color= 'darkred' size = '5'>**Python Tuple**</font>
    
A tuple in Python is similar to a list, but tuples are unchangeble.

Unchangeable Tuples meaning that we cannot **change, add or remove** items after the tuple has been created.

<font color= 'steelblue' size = '4'> **Creating a Tuple**</font><br>

A tuple is created by placing all the items (elements) inside parentheses (), separated by commas. The parentheses are optional, however, it is a good practice to use them.(we can also create tuples without using parentheses)

A tuple can have any number of items and they may be of different types (integer, float, list, string, etc.).



In [75]:
# Different types of tuples

# Empty tuple
my_tuple = ()
print(my_tuple)

my_tuple = (1, 2, 3)
print(my_tuple)

# tuple with mixed datatypes
my_tuple = (1, "Hello", 3.4)
print(my_tuple)

# nested tuple
my_tuple = ("mouse", [8, 4, 6], (1, 2, 3))
print(my_tuple)

#without paranteses

my_tuple = 'reez',1,2,3,'goofy'
print(my_tuple)
print(len(my_tuple))
print(my_tuple[0])
#my_tuple[1] = 'art' #error


()
(1, 2, 3)
(1, 'Hello', 3.4)
('mouse', [8, 4, 6], (1, 2, 3))
('reez', 1, 2, 3, 'goofy')
5
reez


<font color ='steelblue' size ='4'>**Create a Python Tuple With one Element**</font>

In Python, creating a tuple with one element is a bit tricky. Having one element within parentheses is not enough.

We will need a trailing comma to indicate that it is a tuple,


In [21]:
var1 = ("Hello") # string
var2 = ("Hello",) # tuple

In [23]:
mytuple = ("apple", "banana", "cherry")
print(type(mytuple))
thislist = tuple (["apple", "banana", "cherry"])
print(type(thislist))

<class 'tuple'>
<class 'tuple'>


Like a list, each element of a tuple is represented by index numbers (0, 1, ...) where the first element is at index 0.

We use the index number to access tuple elements. For example:

<span style = 'color:steelblue;font-weight:600;font-size:20px'>Indexing and slicing</span><br>
We can use the index operator [] to access an item in a tuple, where the index starts from 0.

So, a tuple having 6 elements will have indices from 0 to 5. Trying to access an index outside of the tuple index range( 6,7,... in this example) will raise an IndexError.

The index must be an integer, so we cannot use float or other types. This will result in TypeError

In [91]:
letters = ("p", "r", "o", "g", "r", "a", "m", "i", "z")

print(letters[0])   
print(letters[5])
print(letters[-2])
print(letters.index('z')) #location 

print(letters[1:4])
print(letters[:-7])

p
a
i
8
('r', 'o', 'g')
('p', 'r')


In [112]:
#what i mean by unchangable #we cant use append فلان 
x = ("apple", "banana", "cherry", "melon", "lemon")
y = list(x)
y.remove("lemon") #use list method
y[1] = "kiwi"
x = tuple(y)
print(x)
print('__________')

# u cant add items in ur tuple but tuple + tuple is true
thistuple = ("apple", "banana", "cherry")
z = ("orange",)
thistuple += z
# .append give us an error for tuple
print(thistuple)

#unpack
car = ('blue','auto', '7mil')
color, model, price = car
print(color)
print(price)

('apple', 'kiwi', 'cherry', 'melon')
__________
('apple', 'banana', 'cherry', 'orange')
blue
7mil


str

<font color= 'steelblue' size = '3'>**Python Tuple Methods**</font><br>

In Python ,methods that add items or remove items are not available with tuple. Only the following two methods are available.

Some examples of Python tuple methods:



In [130]:
my_tuple = ('a', 'p', 'p', 'l', 'e',)
numbs = (9,1,2,3,4,5,2,3,4,5,6,7,0)

print(my_tuple.count('p'))#count
print(numbs.count(9))
print(sum(numbs))
print(max(numbs))
print(min(numbs))
print(tuple(reversed(numbs)))
c = zip(my_tuple, numbs)  #zip
c = list(c)
print(c)
#unzip
u = zip(*c)
print(list(u))

2
1
51
9
0
(0, 7, 6, 5, 4, 3, 2, 5, 4, 3, 2, 1, 9)
[('a', 9), ('p', 1), ('p', 2), ('l', 3), ('e', 4)]
[('a', 'p', 'p', 'l', 'e'), (9, 1, 2, 3, 4)]


<font color = 'steelblue' size = '3' >**Iterating through a Tuple in Python**</font>

We can use the for loop to iterate over the elements of a tuple. For example,

In [12]:
languages = ('Python', 'Swift', 'C++')

# iterating through the tuple
for language in languages:
    print(language)

Python
Swift
C++


In [1]:
mytuple = ('reza','ali','sajad','nvid','kave','mina')
for i in range(len(mytuple)):
#or for i range mytuple:
       #print(i)
    print(mytuple[i])

reza
ali
sajad
nvid
kave
mina


In [152]:
#can you count the tuoles?!
num = [1,3, (8,) , (7,8), 9,6, (42), (34,2),'c'] 
c = 0
for i in num:
    if isinstance(i, tuple):    #methods for finding tuples
       c+= 1
print(c)

3


In [195]:
#can you change 3 and 6 to 9?!
a = [(1,2,3), (4,5,6)]
b = [i[:-1]+(9,) for i in a]
print(b)

[(1, 2, 9), (4, 5, 9)]


<font color= 'steelblue' size = '3'>**Advantages of Tuple over List in Python**</font>

Since tuples are quite similar to lists, both of them are used in similar situations.

However, there are certain advantages of implementing a tuple over a list:

1. We generally use tuples for heterogeneous (different) data types and lists for homogeneous (similar) data types.
2. Since tuples are immutable, iterating through a tuple is faster than with a list. So there is a slight performance boost.
3. Tuples that contain immutable elements can be used as a key for a dictionary. With lists, this is not possible.
4. If you have data that doesn't change, implementing it as tuple will guarantee that it remains write-protected.

<span style ='color:darkred;font-weight:700;font-size:30px'> Python Strings </span>

<font color= 'steelblue' size = '4'>**Methods of Python String**</font>

Besides those mentioned above, there are various string methods present in Python. Here are some of those methods:

| Methods   |	Description |
| :-        |   :-          |
|upper()	|converts the string to uppercase|
|lower()    |converts the string to lowercase
|zfill()    |adds zeros (0) at the beginning
|replace()	|replaces substring inside
|find()	|returns the index of first occurrence of substring
|rstrip()	|removes trailing characters
|split()	|splits string from left
|isdigit()	|checks ndigits characters
|index()	|returns index of substring|

<font color= 'steelblue' size = '4'>**Escape Sequences, strip(), replace(), split()** </font>

- The escape sequence is used to escape some of the characters present inside a string.

- The **split()** method returns a list  where the text between the specified separator becomes the list items.

In [7]:
a = 'abcdef'
print(a[1:4]) #bcd
print(a[::-1]) #barax
print('______________________________')

s = 'python3 strings data'
print(len(s))
print(max(s))           #y
print('th' in s)       #True
print('st' not in s)   #true
print(s.islower())    #true
print(s.isdigit())    #false

print('______________________________')

d = '12a3bcd4'    #jame adad dar string
k = 0
for i in d:
    if i.isdigit()== True:
        k += int(i)
print(k)

print('______________________________')

f = 'welcome to my soul society!'
print(f.startswith('we')) #true
print(f.endswith('ty'))   #false !
print(f.find('o'))        #4
print(f.index('o',5))   #index will give us an error if dont find a character but find wont
print(f.count('e'))
print(f.capitalize())   #Welcome to my soul society!
print(f.title())        #Welcome To My Soul Society!
print(f.upper())        #WELCOME TO MY SOUL SOCIETY!   
print(f.replace('soul society', 'home'))
print(f.split())    # .join yek list ra jomle mikonad baraxe split

print('______________________________')

q = 'zara@gmail.com'
i = q.find('@')    
print(q[i+1:])    #gmail.com
print(q.split('@'))

print('______________________________')

k = '$$$python!$$$'
print(k.strip('$'))    #python!
print(k.lstrip('$'))  #python!$$$
print(k.rstrip('$'))   #$$$python!

ad = '###www.sanjesh.org!$$$$'
print(ad.lstrip('###').rstrip('$$$'))



bcd
fedcba
______________________________
20
y
True
False
True
False
______________________________
10
______________________________
True
False
4
9
3
Welcome to my soul society!
Welcome To My Soul Society!
WELCOME TO MY SOUL SOCIETY!
welcome to my home!
['welcome', 'to', 'my', 'soul', 'society!']
______________________________
gmail.com
['zara', 'gmail.com']
______________________________
python!
python!$$$
$$$python!
www.sanjesh.org!


<font color = 'steelblue' size ='4'>**format(), count(), index(), is digit(), zfill(), join(),translate()** </font>


- we can combine strings and numbers by using the **format()** method!<br>The format() method takes the passed arguments, formats them, and places them in the string where the placeholders {}

- he **index()** method is almost the same as the find() method, the only difference is that the find() method returns -1 if the value is not found, **rindex()** will find the last given value
- **isdigit()** Check if all the characters in the text are digits
- The **zfill()** method adds zeros (0) at the beginning of the string, until it reaches the specified length.<br> if the value of the len parameter is less than the length of the string, no filling is done

In [47]:
age = 36
name = 'mamad'
txt = "My name is {0}, and I am {1}".format(name,age)
print(txt)

n = 16
m = 15
f = 17.87546432
p = 0.833    # %
k = 2000000000000

print('{0:f} {1:d}'.format(n, m))
print('{:.2f}'.format(f))
print('{:.2%}'.format(p))
print('{:,}'.format(k))

My name is mamad, and I am 36
16.000000 15
16.000000 15
17.88
83.30%
2,000,000,000,000


In [7]:
d = 'this new tree is to be here to live long life'
m = '23345'
print(d.count('t'))
print(d.index('h'))
m.isdigit()

4
1


True

In [15]:
a = 345
b = str(a).zfill(6) #zero fill
print(b)

d ='00010'
print(str(int(d)))

s = 'sara'
print(s.ljust(8, '*'))
print(s.rjust(8, '*'))
print(s.center(8, '*'))

000345
10
sara****
****sara
**sara**


- The **translate()** method returns a string where some specified characters are replaced with the character described in a dictionary, or in a mapping table.
- Use the **maketrans()** method to create a mapping table.
- The **join()** method takes all items in an iterable and joins them into one string. A string must be specified as the separator.

In [34]:
a = ' this life is full of shit'
b = str.maketrans('l', 'r')
print(a.translate(b))


 this rife is furr of shit


<font color= 'darkred' size = '5'>**Pyhton Set**</font>

<font color= 'steelblue' size = '3'>**Create a Set in Python**</font>

In Python, we create sets by placing all the elements inside curly braces {}, separated by comma.

A set can have any number of items and they may be of different types (integer, float, tuple, string etc.).<br>But a set cannot have mutable elements like lists, sets or dictionaries as its elements.<br> sets are unchangable but we can add and remove new items.


<font color='steelblue'>**Notice that**</font><br> you might get output in a different 
order. This is because the set has no particular order.

|Method	|Description|
|:-     |:-         |
|add()	|Adds an element to the set|
|clear()	|Removes all elements from the set
|copy()	|Returns a copy of the set
|difference()	|Returns the difference of two or more sets as a new set
|discard()	|Removes an element from the set if it is a member. (Do nothing if the element is not in set)
|intersection()	|Returns the intersection of two sets as a new set
|issubset()	|Returns True if another set contains this set
|issuperset()	|Returns True if this set contains another set|

In [14]:
# create a set of integer type
student_id = {112, 114, 116, 118, 115}
print('Student ID:', student_id)


# create a set of mixed data types

mixed_set = {'Hello', 101, True , 'Bye'}
print('Set of mixed data types:', mixed_set)
len(mixed_set)
type(mixed_set)

mytupleset = set((2,3,4,5,6,7))
print(8 in mytupleset)

f = set(('apple','cherry','mango','banana'))
f.add('cucumber')
print(f)
f.update(['mango','grapes'])
print(f)
f.remove('apple')
f


Student ID: {112, 114, 115, 116, 118}
Set of mixed data types: {'Bye', 'Hello', 101, True}
False
{'mango', 'apple', 'cucumber', 'cherry', 'banana'}
{'mango', 'apple', 'cucumber', 'cherry', 'grapes', 'banana'}


{'banana', 'cherry', 'cucumber', 'grapes', 'mango'}

In [1]:
# create an empty set
empty_set = set()

# create an empty dictionary
empty_dictionary = { }

#______________________

numbers = {2, 4, 2, 8, 4, 6, 6, 2, 8}
print(numbers)

#Here, we can see there are no duplicate items in the set as a set cannot contain duplicates.



{8, 2, 4, 6}


- Sets are **mutable**. However, since they are unordered, **indexing** has no meaning.

- We cannot access or change an element of a set using indexing or slicing. Set data type does not support it.

- we use the **add()** method to add an item to a set<br>

- The **update()** method is used to update the set with items other collection types (lists, tuples, sets, etc).
- the **union()** method is used to join 2 sets, note that both union() and update() will exclude any duplicate items.

- We use the **discard()** or **remove()** method to remove the specified element from a set

- You can also use the **pop()** method to remove an item, but this method will remove a random item, so you cannot be sure what item that gets removed.
- also **clear()** and **del** will remove whole set

In [55]:
numbers = {21, 34, 54, 12}

print('Initial Set:',numbers)

# using add() method
numbers.add(32)

print('Updated Set:', numbers)
#_____________________________________

companies = {'Lacoste', 'Ralph Lauren'}
tech_companies = ['apple', 'google', 'apple']
tech1 = ('sumsung',)
companies.update(tech_companies)  #{'apple', 'google', 'Lacoste', 'Ralph Lauren', 'sumsung'}
companies.update(tech1)
print(companies)
print('_________________')
#DISCARD AND REMOVE

languages = {'Swift', 'Java', 'Python', 'R', 'perl'}
languages.discard('R') 
print(languages)

languages.discard('GO') #no error but remove will raise an error
print(languages)
#remove()
languages.remove('Swift')
print(languages)

print('_________________')

#pop() #randomly remove

thisset = {"apple", "banana", "cherry"}
x = thisset.pop()

print(x)
print(thisset)
print('_________________')
c = companies.copy()
print(c)
c.clear()
c

Initial Set: {34, 12, 21, 54}
Updated Set: {32, 34, 12, 21, 54}
{'apple', 'Lacoste', 'Ralph Lauren', 'sumsung', 'google'}
_________________
{'perl', 'Java', 'Swift', 'Python'}
{'perl', 'Java', 'Swift', 'Python'}
{'perl', 'Java', 'Python'}
_________________
cherry
{'apple', 'banana'}
_________________
{'apple', 'Lacoste', 'sumsung', 'Ralph Lauren', 'google'}


set()

<font color= 'steelblue' size = '3'>**Python Set Operations**</font><br>

Python Set provides different built-in methods to perform mathematical set operations like union, intersection, subtraction, and symmetric difference.

1. The union of two sets A and B include all the elements of set A and B.<br> we use the **|** operator or the **union()** method to perform the set union operation<br> and it is = A ⋃ B set operation.

2. Set Intersection <br> The intersection of two sets A and B include the common elements between set A and B. <br> In Python, we use the **&** operator or the **intersection()** method to perform the set intersection operation. 

3. Difference between Two Sets <br> The difference between two sets A and B include elements of set A that are not present on set B.<br> We use the **-** operator or the **difference()** method to perform the difference between two sets.
4. Set Symmetric Difference <br> The symmetric difference between two sets A and B includes all elements of A and B without the common elements.<br> In Python, we use the **^** operator or the **symmetric_difference()** method to perform symmetric difference between two sets

In [75]:
a = {1,2,3,4,5}
b= {1,3,5,7,9}
print(a-b)
print(b-a)
r= a.difference(b) #tafazol
print(r)
print(a,b)
r= a.difference_update(b) #update a
print(a,b)
a = {1,2,3}
b = {2,3,4}
print(a.symmetric_difference(b)) #tafazol moteqaren
print(a^b)
print(a.intersection(b)) #eshterak
print(a&b)
print(a.union(b))
print(a|b)

{2, 4}
{9, 7}
{2, 4}
{1, 2, 3, 4, 5} {1, 3, 5, 7, 9}
{2, 4} {1, 3, 5, 7, 9}
{1, 4}
{1, 4}
{2, 3}
{2, 3}
{1, 2, 3, 4}
{1, 2, 3, 4}


In [88]:
a = {1,2}
b = {1,2,3}
print(a.isdisjoint(b)) #aya mostaqelan do set :NO
print(a.issubset(b)) #zir majmOe

w = 'alireza'
x = {'a', 'r', 's'}
s = set(w)
print(x.intersection(s))

False
True
{'r', 'a'}


<font color= 'steelblue' size= '3'>**Built-in Functions with Set**</font>


|Function	|Description|
|:-         |:-         |
|all()	    |Returns True if all elements of the set are true (or if the set is empty).
|any()	|Returns True if any element of the set is true. If the set is empty, returns False.
|enumerate()	|Returns an enumerate object. It contains the index and value for all the items of the set as a pair.
|len()	|Returns the length (the number of items) in the set.
|max()	|Returns the largest item in the set.
|min()	|Returns the smallest item in the set.
|sorted()	|Returns a new sorted list from elements in the set(does not sort the set itself).
|sum()	|Returns the sum of all elements in the set|

<font color= 'darkred' size = '5'>**Python dictionary**</font>

Dictionary items are ordered, changeable, and does not allow duplicates(same key)

Dictionary items are presented in key:value pairs, and can be referred to by using the key name.

As of Python version 3.7, dictionaries are ordered. In Python 3.6 and earlier, dictionaries are unordered.<br> 
When we say that dictionaries are **ordered**, it means that the items have a defined order, and that order will not change.<br>
Unordered means that the items does not have a defined order, you cannot refer to an item by using an index.

Dictionaries are **changeable**, meaning that we can change, add or remove items after the dictionary has been created.

In [40]:
thisdict = {
  "brand": "dodge",
  "model": "challenger",
  "year": '1968'
}
print(thisdict["brand"])
print(thisdict.get('model')) #method call
print(len(thisdict))
print(thisdict.keys()) #method
print(thisdict.values())
print('_____________')

d= dict(brand = 'ford',model = 'mustang', color = 'navy')
d['year'] = '1989'
print(d)
print(d['model']) 

d['model'] = 'branco' #changeable
print(d.get('model'))
print(d) 
print(list(d.items()))


dodge
challenger
3
dict_keys(['brand', 'model', 'year'])
dict_values(['dodge', 'challenger', '1968'])
_____________
{'brand': 'ford', 'model': 'mustang', 'color': 'navy', 'year': '1989'}
mustang
branco
{'brand': 'ford', 'model': 'branco', 'color': 'navy', 'year': '1989'}
[('brand', 'ford'), ('model', 'branco'), ('color', 'navy'), ('year', '1989')]


<font  color='darkred' size = '4'>**Some dictionary functions**</font>

|Method    | description      |
| :-       | :-               |
|clear()   |Removes all the elements|
|copy()	   |Returns a copy of the dictionary
|fromkeys()|Returns a dictionary with the specified keys and value
|get()	    |Returns the value of the specified key
|items()	|Returns a list containing a tuple for each key value pair
|keys()	    |Returns a list containing the dictionary's keys
|values()	|Returns a list of all the values in the dictionary
|pop()	    |Removes the element with the specified key
|popitem()	|Removes the last inserted key-value pair
|update()   |Updates the dictionary with the specified key-value pairs|

In [8]:
myDict = {"year": input('year: '), "month": input('month: ') , 'date': input('date: ')}

x = '/'.join(myDict.values())

print('date of birth: ',x)

year: 1998
month: 11
date: 06
date of birth:  1998/11/06


In [56]:
car = {
"brand": "dodge",
"model": "charger",
"year": 1968
}
car.pop('model') #remove 
print(car)

car.popitem() #remove an item
print(car) 

car.clear()  #empty dict
print(car)

{'brand': 'dodge', 'year': 1968}
{'brand': 'dodge'}
{}


In [62]:
a = [ 'x' , 'y' , 'x', 'z', 'y' ,'x']
d = {}
for i in a:
    if i not in d:
        d[i] = 1
    else:
        d[i]+=1
print(d)

{'x': 3, 'y': 2, 'z': 1}


<span style="color:steelblue;font-weight:300;font-size:16px"> **Change, add, remove items**  </span>

In [66]:
mydictionary = {'car' : 'ford', 'model' : 'mustang' , 'year' : '1998','motor' : 'a_alpha' , 'class' : 'c'}
mydictionary.update({'color' :'red'})
#mydictionary.update({'model' :'safari'}) #change

del mydictionary['class']
                 
#mydictionary.clear() #empty whole dictionary
mydictionary

{'car': 'ford', 'model': 'mustang', 'year': '1998', 'color': 'red'}

In [68]:
for x, y in mydictionary.items():
    print(x,':', y)

car : ford
model : mustang
year : 1998
color : red


In [30]:
for x, y in mydictionary.items():
    print(x,':', y)

car : ford
model : mustang
year : 1998
color : red


<font color = 'steelblue' size = '3'> **copy a dictionary**</font>

In [31]:
thisdict = {
  "brand": "Ford",
  "model": "Mustang",
  "year": 1964
}
mydict = thisdict.copy()
print(mydict)
#or 
#mydict = dict(thisdict)
#print(mydict)

{'brand': 'Ford', 'model': 'Mustang', 'year': 1964}


In [92]:
s = 'anddjsanadowwnarafnf'
d = {}
for i in s:
    if i not in d:
        d[i] = 1
    else:
        d[i] +=1        
print(d)

#or
s = 'anddjsanadowwnarafnf'
d = {}
for i in s:
    d[i] = d.get(i, 0) + 1
print(d)
print('_____________________')

line = 'a dictionary is a data structure.'
d={}
s = line.split()
for i in s:
    d[i] = d.get(i, 0) +1
print(d)

{'a': 5, 'n': 4, 'd': 3, 'j': 1, 's': 1, 'o': 1, 'w': 2, 'r': 1, 'f': 2}
{'a': 5, 'n': 4, 'd': 3, 'j': 1, 's': 1, 'o': 1, 'w': 2, 'r': 1, 'f': 2}
_____________________
{'a': 2, 'dictionary': 1, 'is': 1, 'data': 1, 'structure.': 1}


In [116]:
#sumation of numbers in dict

d = {'a': 5, 'n': 4, 'd': 3, 'j': 1, 's': 1, 'o': 1, 'w': 2, 'r': 1, 'f': 2}
s = 0
for i in d:
    s += d[i]
print(s)
#or
print(sum(d.values()))

20
20


In [120]:
#sort
d = {'a': 5, 'n': 4, 'd': 3, 'j': 1, 's': 1, 'o': 1, 'w': 2, 'r': 1, 'f': 2}

import operator 
k = operator.itemgetter(1) #reverse
print(sorted(d.items(), key = k))

k = operator.itemgetter(0)
print(sorted(d.items(), key = k))

print()

scores = {'ali' :[16,14,12], 'reza' : [16,19,17], 'kaveh' :[10, 17,12]}

d = {k: sorted(v) for k,v in scores.items()}
print(d)

[('j', 1), ('s', 1), ('o', 1), ('r', 1), ('w', 2), ('f', 2), ('d', 3), ('n', 4), ('a', 5)]
[('a', 5), ('d', 3), ('f', 2), ('j', 1), ('n', 4), ('o', 1), ('r', 1), ('s', 1), ('w', 2)]

{'ali': [12, 14, 16], 'reza': [16, 17, 19], 'kaveh': [10, 12, 17]}


In [9]:
#merging 2 dictionary
d1 = {'a': 5, 'n': 4, 'd': 3, 'j': 1}
d2 = {'s': 1, 'o': 1, 'w': 2, 'r': 1}
d = {}
d = d1.copy()
d.update(d2)
print(d)
#or 
d={}
for i in (d1,d2):
    d.update(i)
print(d)


{'a': 5, 'n': 4, 'd': 3, 'j': 1, 's': 1, 'o': 1, 'w': 2, 'r': 1}
{'a': 5, 'n': 4, 'd': 3, 'j': 1, 's': 1, 'o': 1, 'w': 2, 'r': 1}


In [25]:
k = ['red', 'green']
v = ['#FF0000','#008000']

z = zip(k,v)
d = dict(z)
print(d)

{'red': '#FF0000', 'green': '#008000'}


In [27]:
s = 'alireza'
x = ['a','r']
d ={}
for i in s:
    if i in x:
        d.setdefault(i,0)
        d[i] +=1
print(d)

{'a': 2, 'r': 1}


In [73]:
#removing duplicated values
m = {'a': 5, 'n': 4, 'd': 3, 'j': 4, 't': 5, 'm': 8, 'p': 3}
r ={}

for k,v in m.items():
    if v not in r.values():
        r[k] = v
print(r)
        

{'a': 5, 'n': 4, 'd': 3, 'm': 8}


In [95]:
d = {'head' :0, 'tail':0}
s = list(d.keys())
import random
for i in range(10):  #10 bar seke ra part mikonim
    d [random.choice(s)] +=1
    
print(d)

{'head': 2, 'tail': 8}
